### PR TITLE
[Graphbolt] Enhance preprocessing: Ensure relative paths in YAML and convert to absolute paths before loading.

### PR DESCRIPTION
**Pull request by keli-wen**
## Description

This PR primarily addresses the issue: [[Graphbolt] Update OndiskDataset loading logic](https://github.com/dmlc/dgl/issues/6022).

In the previous PR: [[Graphbolt] Update preprocess to Accept Dataset Path Instead of YAML File Path](https://github.com/dmlc/dgl/pull/6091). We've observed that moving the `dataset` directory invalidates the prior preprocessing due to absolute path issues.

To address this, we now ensure that after preprocessing, all paths within the YAML file are relative. To ensure data accessibility from any location, we've introduced a `_convert_yaml_path_to_absolute_path` function, which temporarily converts paths in the YAML to absolute paths before loading.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by frozenbugs — APPROVED** [review]

**Comment by frozenbugs** [comment]
No test change?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 1]
```diff
@@ -274,10 +278,12 @@ class OnDiskDataset(Dataset):
     def __init__(self, path: str) -> None:
         # Always call the preprocess function first. If already preprocessed,
         # the function will return the original path directly.
+        self.dataset_dir = path
         path = preprocess_ondisk_dataset(path)
         with open(path) as f:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:283]:
```suggestion
        yaml_data = preprocess_ondisk_dataset(path)
        with open(yaml_data) as f:
```

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 1 above)
Comment (keli-wen) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:346]: the `preprocessed_ondisk_dataset` return the `path` of preprocessed metadata.yaml

Comment (frozenbugs) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:346]: Oh, sorry, I intended to rename it to yaml_path.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 1 above)
Comment (keli-wen) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:346]: I'll rename it in next `load` PR～

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 2]
```diff
@@ -81,7 +81,7 @@ def preprocess_ondisk_dataset(dataset_dir: str) -> str:
         return os.path.join(dataset_dir, "preprocessed/metadata.yaml")
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:81]: just curious, does this work for windows?

File: tests/python/pytorch/graphbolt/test_ondisk_dataset.py [hunk 3]
```diff
@@ -1063,3 +1062,144 @@ def test_OnDiskDataset_preprocess_path():
             match=r"metadata.yaml does not exist.",
         ):
             _ = gb.OnDiskDataset(fake_dir)
+
+
+def test_OnDiskDataset_preprocess_yaml_content():
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_ondisk_dataset.py:1067]: Can you add test for windows?

**Comment by frozenbugs** [comment]
Excellent, the test is way cleaner now.

**Review by Rhett-Ying — APPROVED** [review]

**Pull request merged**